[FEAT] 푸시 알림 기능

kimdevv opened the pull request:
## Issue Number
closed #1160 

## As-Is
* 채팅을 수신해도 알림이 뜨지 않아 수시로 확인해야 하는 상황이었습니다.
* 원활한 채팅을 위해서는 채팅방에서 나가지 않고 상시 대기하고 있어야 하는 상황이었습니다.

## To-Be
* FCM 토큰을 저장하는 API를 추가하였습니다.
* FCM 토큰을 통해 메세지(푸시 알림)를 전송하는 기능을 구현하였습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
<img width="382" height="107" alt="image" src="https://github.com/user-attachments/assets/a19016a8-b400-4e3e-bfaa-8cda4a48f1ef" />

<img width="1280" height="757" alt="image" src="https://github.com/user-attachments/assets/de3c6c3c-1527-4484-8cbd-2f135872fbc2" />
위와 같은 흐름으로 일단 기능 자체만 간단하게 구현하였습니다. 제가 생각했을 때 현재 아래와 같은 단점이 있습니다.

- 외부 네트워크 호출 작업이 동기적으로 이루어지고 있다. 따라서 네트워크 성능에 따라 푸시 알림 발송 기능의 성능이 느려질 수 있다. 푸시 알림 발송 요청이 한 번에 많이 들어오면 커넥션 풀 고갈 위험도 있다.
- 재시도 로직이 없다. 푸시 발송이 실패하면 다시 전송하지 않고 커스텀 예외를 던진 채 작업이 종료된다.

SQS 및 람다를 활용하여 푸시 알림 요청을 비동기적으로 처리하고, DLQ를 도입해 푸시 발송 실패 시 재시도하도록 리팩터링하려고 합니다. 더 좋은 방법이 있다면 제안해 주시면 감사하겠습니다.



구현하면서 제가 가졌던 생각의 흐름은 [여기](https://kimdev-s.tistory.com/367)를 보시면 됩니다.

kimdevv committed:
feat: 회원 FCM 토큰 테이블 및 엔티티 추가

- member_fcm_token 테이블 생성(migration 스크립트 추가)
  - id(PK), member_id(FK), token, updated_at 컬럼 구성
  - token에 유니크 제약 조건 추가
  - member_id에 인덱스 추가
- MemberFcmToken 엔티티 추가
  - Auditing을 통한 updatedAt 자동 업데이트 설정
  - 데이터 엔티티의 일관성과 보안을 위해 Lombok 접근 제어자 활용

kimdevv committed:
feat: 회원 FCM 토큰 업서트 기능 추가

- FcmTokenUpsertRequest DTO 추가
- MemberFcmTokenRepository CRUD 인터페이스 정의
- MemberFcmTokenService에 FCM 토큰 저장 및 갱신 로직 구현
  - 회원 존재 여부 검증 메서드 추가
  - 신규 토큰 저장/기존 토큰 갱신 기능 추가
- NotificationController에 FCM 토큰 업서트 HTTP API 구현
- MemberFcmToken 엔티티에 생성자 및 토큰 갱신 메서드 추가

kimdevv committed:
test: 회원 FCM 토큰 서비스 테스트 코드 추가

- FCM 토큰 업서트 로직에 대한 단위 테스트 케이스 작성
  - 신규 토큰 저장 테스트
  - 기존 토큰 갱신 테스트
  - 존재하지 않는 회원 요청 예외 처리 테스트
- MemberFcmToken 엔티티에 token 필드 Getter 추가

kimdevv committed:
feat: FCM 알림 전송 기능 추가

- FCM 전송을 위한 NotificationService 인터페이스 정의
- Firebase 초기화 및 인증을 위한 FcmConfiguration 클래스 추가
- FcmNotificationService 구현체에 FCM 알림 전송 로직 구현
  - 회원의 FCM 토큰 조회 및 알림 전송 로직 포함
  - FirebaseMessaging 객체 활용
- 예외 처리 클래스 추가 (MemberFcmTokenNotFoundException, FcmSendException)
- InfraErrorMessage, BusinessErrorMessage 메시지 항목 추가
- build.gradle에 Firebase Admin SDK 종속성 추가
- NotificationController에 알림 관련 API RequestMapping 경로 리팩토링

kimdevv committed:
test: FCM 토큰 업서트 통합 테스트 추가

- 신규 FCM 토큰 저장 통합 테스트 작성
- 기존 FCM 토큰 갱신 통합 테스트 작성
- 존재하지 않는 회원 예외 처리 통합 테스트 작성

kimdevv committed:
fix: FCM 알림 전송 메서드 매개변수 순서 변경

kimdevv commented on backend/fittoring/src/main/java/fittoring/domain/model/FcmToken.java:
```diff
@@ -0,0 +1,47 @@
+    private Long id;
+
+    @Column(name = "member_id", nullable = false, unique = true)
+    private Long memberId;
```
멤버 한 명이 여러 대의 기기에서 로그인할 것 같지는 않아서 `unique` 옵션을 걸어 두었습니다. 여러 대의 기기에서 로그인을 허용하게 된다면 `unique` 옵션을 제거하고, `Agent`와 같이 데스크톱, 모바일, 태블릿 등을 구별할 수 있는 Enum값이 필요할 것 같습니다. 정책상 논의 내용이라 추후 이야기해 봐야 할 것 같아요.

kimdevv replied:
그리고 Fcm 토큰과 memberId를 가지고 있게 해서, 로그인되지 않은 유저에게는 푸시 알림을 보낼 수 없는 상황입니다. 더 나은 설계 방법이 없는 것 같아서 일단 현재 구조를 차용했는데, 좋은 내용 있다면 제안 부탁드립니다.

jye326 replied:
1. 멤버 한 명이 여러 대의 기기에서 로그인하는 상황은 충분히 있을 것 같습니다. 데스크톱으로 처음 서비스를 접한 유저가 모바일로 시도해 볼 수도 있고, 친구의 핸드폰으로 저희 서비스를 시연해 볼 수도 있을 것 같아요. 그래서 `unique` 는 빼고 `isActivated`등의 디바이스 활성화 여부 속성을 추가해서 현재 접속한 기기를 판단하는 것은 어떨까요? 로그인 시에 해당 디바이스를 제외한 다른 토큰의 activate를 false로 변경하는 거죠. 어떻게 생각하시나요?
2. 로그인 하지 않은 유저에게 푸시 알림을 보낼 필요는 없을 것 같아요.

dye0p replied:
> 로그인되지 않은 유저에게는 푸시 알림을 보낼 수 없는 상황입니다. 더 나은 설계 방법이 없는 것 같아서 일단 현재 구조를 차용했는데, 좋은 내용 있다면 제안 부탁드립니다.

`member_id`를 null 허용으로 변경하고, `device_id` 같은 컬럼을 추가해서 기기별로 토큰 식별을 한다면 어떨까요? `member_id`는 로그인이 성공하면 그때 업데이트 해줄 수 있을 것 같아요.

kimdevv commented on backend/fittoring/src/main/java/fittoring/domain/model/MemberFcmToken.java:
```diff
@@ -0,0 +1,47 @@
+@EntityListeners(AuditingEntityListener.class)
+@Table(name = "member_fcm_token")
+@Entity
+public class MemberFcmToken {
```
`RefreshToken`과 마찬가지로 소프트 삭제가 크게 필요 없을 것 같아서 적용하지 않았습니다.
* 삭제 내역을 보관할 필요가 없는 엔티티라고 판단함.
* FCM 토큰은 주기적으로 갱신이 일어나기 때문에, 소프트 삭제 시 테이블에 삭제된 레코드들이 많이 쌓일 것 같음.

dye0p replied:
p3: 토큰 삭제는 어느 주기로 일어나나요?

jye326 replied:
동의합니다.

kimdevv replied:
아직 삭제는 이루어지지 않습니다. 리프레쉬 토큰이랑 유사합니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/infrastructure/FcmNotificationSender.java:
```diff
@@ -0,0 +1,45 @@
+                        .setBody(body)
+                        .build())
+                .build();
+        try {
+            FirebaseMessaging.getInstance().send(message);
+        } catch (FirebaseMessagingException exception) {
+            throw new FcmSendException(InfraErrorMessage.FCM_SEND_ERROR + exception.getMessage());
+        }
```
플랫폼인만큼 관리 및 디버깅을 위해 푸시 알림 발송 내역을 저장하는 테이블도 있으면 좋을 것 같아요. (member_id, token, isSucess, title, body) 정도...? 동의하신다면 추후에 만들어 보겠습니다.

jye326 replied:
p3 : 내역을 남기는 것은 동의합니다만 로깅과 내용이 중복되지 않을까 싶네요.

kimdevv replied:
로깅은 API 요청 및 응답 내역을 저장하고 있는 반면 푸시 알림 내역에는 푸시 알림 내용이랑 알림 수신자, 알림 발송 성공 유무 등을 저장할 것 같아요

dye0p replied:
좋은데요? 필요할 것 같아요 동의합니다~

kimdevv commented on backend/fittoring/src/main/java/fittoring/config/FcmConfiguration.java:
```diff
@@ -0,0 +1,33 @@
+                FirebaseApp.initializeApp(options);
+            }
+        } catch (Exception exception) {
+            throw new RuntimeException("Firebase 초기화 실패: " + exception.getMessage());
```
어떤 예외를 던져야 할 지 모호해서 `RuntimeException`을 던졌습니다. FCM을 초기화하는 부분에서 예외가 발생하는 상황은 어떻게 보면 비즈니스 예외, 어떻게 보면 서버측 예외로도 보이는 거 같아서... 의견 부탁드립니다. 지금 PR을 작성하면서 느끼는 건데 이 부분은 인프라 관련된 예외니까 서버 오류를 던지는 게 맞는 것 같기도 하네요.

kimdevv committed:
Merge branch 'develop' into feature/1160-push
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

kimdevv committed:
feat: FCM 설정에 프로파일 추가

- FcmConfiguration 클래스에 @Profile 어노테이션 추가
- dev, prod 환경에서만 FCM 설정 활성화되도록 수정

kimdevv committed:
feat: 채팅 메시지 등록 시 알림 전송 기능 추가

- ChatMessageService에 NotificationService 의존성 주입
- 메시지 등록 시 상대방에게 알림 전송 로직 추가
- ChatRoom에 상대방 ID 조회 메서드 추가 (`getOpponentIdOf`)

jye326 commented on backend/fittoring/src/main/java/fittoring/application/notification/service/MemberFcmTokenService.java:
```diff
@@ -0,0 +1,46 @@
+            saveNewFcmToken(memberId, token);
+            return;
+        }
+        renewFcmToken(tokenOptional.get(), token);
```
- 분기문을 람다로 깔끔하게 리팩터링 할 수 있을 것 같아요!
- `@Transactional` 걸려있으니 JPA 더티체킹으로 `save(fcmToken)`를 명시적으로 호출하지 않아도 될 것 같아요!
```suggestion
        memberFcmTokenRepository.findByMemberId(memberId).ifPresentOrElse(
                foundToken -> foundToken.updateToken(token),    // 존재하면 업데이트
                () -> saveNewFcmToken(memberId, token)  // 없으면 새로 발급
        );
```

kimdevv replied:
람다로 처리할 수 있지만 분기문을 사용하는 쪽이 가독성이 더 좋은 것 같아서 분기문을 사용하였습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/domain/model/MemberFcmToken.java:
```diff
@@ -0,0 +1,37 @@
+@EntityListeners(AuditingEntityListener.class)
+@Table(name = "member_fcm_token")
+@Entity
+public class MemberFcmToken {
```
p2 : 굳이 Entity, 테이블 명 앞에 `member`를 붙일 필요가 없어 보여요. 메서드에서도 `FcmToken`으로 지칭하고 있고, `member`외에 `FcmToken`을 활용할 주체가 없으니 그냥 `FcmToken`으로 통일하는 건 어떨까요?

kimdevv replied:
그렇군요. `Member`은 제외해도 괜찮을 것 같네요!

dye0p commented on backend/fittoring/src/main/java/fittoring/domain/model/Device.java:
```diff
@@ -0,0 +1,37 @@
+    private Long id;
+
+    @Column(name = "member_id", nullable = false, unique = true)
+    private Long memberId;
```
p2: memberId를 간접 참조로 맺은 이유가 궁금합니다! FCM과 Member 사이의 생명주기가 동일 할 것으로 생각되서, 직접 참조를 해도 좋지 않을까 생각이 들어요!

kimdevv replied:
직접 참조도 가능하지만 이전에 앞으로 관계는 간접 참조로 해보자고 했던 기억이 생각나서 간접 참조를 적용했습니다. 또한 알림 기능은 서비스의 규모가 커졌을 때 분리되어야할 1순위 도메인이라고 생각했고, 확장성을 위해 간접 참조를 적용했습니다.

dye0p replied:
JPA가 제공하는 기능을 제대로 사용하지 못할수도 있어서 장단점을 생각해보면 좋을 것 같아요! 지금이 나쁘다는 것은 아닙니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/notification/presentation/NotificationController.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final MemberFcmTokenService memberFcmTokenService;
+
+    @PostMapping("/token")
```
p2: 복수형으로 표현하는 것이 좋을 것 같아요! 
```suggestion
    @PostMapping("/tokens")
```

kimdevv replied:
그렇네요 감사합니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/notification/presentation/NotificationController.java:
```diff
@@ -0,0 +1,26 @@
+    private final MemberFcmTokenService memberFcmTokenService;
+
+    @PostMapping("/token")
+    public ResponseEntity<Void> upsertFcmToken(@RequestBody FcmTokenUpsertRequest requestBody) {
```
p3: requestBody 대신 request는 어떨까요~

kimdevv replied:
Request는 Body를 포함하여 Protocol, Header 등 요청 전반을 포함하는 단어입니다. Dto는 요청 바디를 나타내는 클래스이므로 `requestBody`를 사용하였습니다.

dye0p commented on backend/fittoring/src/main/java/fittoring/config/FcmConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+            if (FirebaseApp.getApps().isEmpty()) {
+                FirebaseApp.initializeApp(options);
+            }
+        } catch (Exception exception) {
```
p2: catch 하고 있는 예외의 범위가 try 구문에서 예상하지 못한 예외도 잡혀버릴 수 있을 것 같습니다. 
`GoogleCredentials.fromStream(resource.getInputStream())` 부분에서 내부적으로 `IOException`을 던지고 있습니다. 외부에서 `IOException`으로 catch 하는것은 어떨까요?

<img width="646" height="155" alt="Image" src="https://github.com/user-attachments/assets/443a9895-52bf-4bbd-a4a5-48bfdfc44aa5" />

kimdevv replied:
FCM 초기화 과정에서 `IOException`만을 처리하게 되면 그 외의 예외는 처리하지 못하게 됩니다. 그러면 `IOException`을 제외한 다른 예외들에 대해서는 `서버에서 심각한 예외가 발생하였습니다.`라는 메세지만 받고 어떤 예외가 발생했는지도 모른 채 디버깅을 진행해야 합니다.
이러한 관점에서 봤을 때 FCM 초기화 과정에서 발생하는 모든 예외를 잡아서 `FCM 초기화 실패 : 원인` 양식으로 처리하는 게 디버깅할 때 더 효율적일 것이라 생각했습니다!

dye0p replied:
> 그 외의 예외는 처리하지 못하게 됩니다

그럼 구체적으로 발생할 수 있는 예외를 catch 해주면 어떨까요? catch는 여러개 해줄 수 있으니 그게 좋을 것 같은데, 어떻게 생각하시나요? Exception은 예외의 범위가 너무 큰 것 같아요!

dye0p reviewed (changes requested):
우선 1차 코멘트 남깁니다! 확인 부탁드려요~

jye326 commented on backend/fittoring/src/main/java/fittoring/application/chat/service/ChatMessageService.java:
```diff
@@ -32,6 +34,9 @@ public ChatMessageResponse registerMessage(Long chatRoomId, ChatMessageRequest r
         chatMessageRepository.save(chatMessage);
+
+        Long opponentId = chatRoom.getOpponentIdOf(senderId);
+        notificationService.sendNotification(opponentId, "핏토링", "채팅이 도착하였습니다.");
```
p2 : 모든 채팅에 대해 푸시 알림이 가게 될 것 같아요! 심지어 현재 화면을 보고 채팅을 하고 있어도 계속 푸시 알림을 받게 될 것 같아요! 푸시 알림 타이밍을 고민해 봐야 할 것 같아요!

kimdevv replied:
모든 채팅에 대해 알림 요청이 전송되긴 하지만, 푸시 알림은 앱이 백그라운드 상태일 때에만 나타납니다! 포그라운드 상태에서는 프론트엔드 단에서 뭔가 처리를 해 주어야 푸시 알림이 나타나기 때문에, 프론트엔드에서 관련 코드를 추가하지 않는다면 괜찮을 것 같아요

jye326 reviewed (commented):
간단히 코멘트 남겼습니다~ 확인 후 재요청주세요!

kimdevv committed:
Merge branch 'develop' into feature/1160-push

kimdevv committed:
fix: FCM 토큰 업서트 API 경로 수정

- 기존 `/token` 경로를 `/tokens`로 변경
- 경로 이름 명확성을 위해 복수형으로 수정

kimdevv committed:
refactor: FCM 토큰 관련 클래스 및 서비스 명칭 수정

- MemberFcmToken을 FcmToken으로 변경
- 관련 예외 및 리포지토리, 서비스, 컨트롤러 클래스 이름 변경
- 각 클래스 내 변수, 메서드, 생성자에서 명칭 변경
- SQL 파일의 테이블 명칭 수정 (member_fcm_token → fcm_token)
- BusinessErrorMessage 메시지 항목 수정 (MEMBER_FCM_TOKEN_NOT_FOUND → FCM_TOKEN_NOT_FOUND)

kimdevv committed:
feat: FCM 토큰 활성화 여부 필드 추가

- FcmToken 엔티티에 `isEnabled` 필드 추가 및 Getter 생성
- 생성자 매개변수에 `isEnabled` 추가
- 알림 전송 시 `isEnabled` 여부에 따라 실행되도록 수정
- FCM 토큰 저장 로직에 기본 활성화 상태 적용
- SQL 스키마에 `is_enabled` 컬럼 추가
- FCM 업서트 API 경로를 `/tokens`로로 수정하여 명확성 향상

jye326 commented on backend/fittoring/src/main/java/fittoring/domain/model/Device.java:
p2:
- 엔티티/테이블 명을 `FcmToken`이 아닌 `Device`로 변경하면 어떨까요? FCM 알림 발송/ 토큰 발급의 기본 단위는 디바이스라고 생각합니다. 그렇게 되면 필드에 token이 있는 것도 더 자연스러울 것 같아요. 지금은 `FcmToken.token`이라 의미가 중복된다고 생각합니다! 관례적으로도 1인 1기기 정책이 아주 강한 경우가 아니면 범용적으로 `Device`를 사용한다고 하네요
- `memberId`에 `unique` 옵션을 해제해야 할 것 같아요!

jye326 committed:
feat: 디바이스 관리 및 FCM 토큰 업서트 로직 리팩토링

- `Device` 엔티티로 리네이밍 및 테이블 구조 변경
- `FcmTokenService`를 `FcmService`로 리팩토링
- `FcmTokenRepository`를 `DeviceRepository`로 교체
- FCM 토큰 등록 및 갱신 로직 수정
- 테스트 케이스 및 관련 유효성 검증 로직 수정
- `FcmNotificationService` 내 로직 변경 및 호환성 추가

dye0p commented on backend/fittoring/src/main/java/fittoring/application/notification/service/FcmService.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private void saveNewFcmToken(Long memberId, String token) {
+        validateMemberExists(memberId);
```
p2: 메서드의 네이밍으로 봤을 때 `validateMemberExists` 가 들어있는 것이 조금 어색한 것 같아요! 외부로 빼내는 것은 어떨까요?

jye326 replied:
리팩터링 되었습니다~

dye0p commented on backend/fittoring/src/main/java/fittoring/domain/model/ChatRoom.java:
```diff
@@ -79,4 +80,12 @@ public ChatRoom(Long reservationId, Long menteeId, Long mentorId) {
+            return mentorId;
+        } else {
+            return menteeId;
+        }
```
p2: else 없이 바로 return 해도 될 것 같아요!

dye0p commented on backend/fittoring/src/main/java/fittoring/infrastructure/FcmNotificationService.java:
```diff
@@ -0,0 +1,47 @@
+                .orElseThrow(() -> new FcmTokenNotFoundException(
+                        BusinessErrorMessage.FCM_TOKEN_NOT_FOUND.getMessage()));
+        if (fcmToken.isEnabled()) {
+            sendNotification(fcmToken.getToken(), title, body);
```
p2: 인프라 계층인 `FcmNotificationService`가 누구에게 보내는지 까지 결정해야 할까요? 누구에게 보낼지 결졍하는 것은 비즈니스 계층이 책임지고, `FcmNotificationService`는 전송에만 신경쓰는것도 좋을 것 같아요! 

테스트시에도 `FcmTokenRepository`를 mocking 해줘야하는 불편함도 있을 것 같습니다.

나아가서 `sendNotification` 메서드는 비즈니스와 인프라 계층 사이의 통신을 담당하는데, `memberId`라는 매개변수를 받아서 member를 찾기보단, token을 찾아서 매개변수로 넘겨주는 것은 어떨까요? 

이렇게 하면 `FcmNotificationService`은 전송의 책임만 질 수 있을 것 같아요!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/chat/service/ChatMessageService.java:
```diff
@@ -24,6 +25,7 @@ public class ChatMessageService {
 
     private final ChatMessageRepository chatMessageRepository;
     private final ChatRoomRepository chatRoomRepository;
+    private final NotificationService notificationService;
```
p2: `NotificationService`를 추상화 한 것은 좋습니다! 다만`FcmNotificationService`가 memberId로 Token을 찾는 비즈니스 로직을 수행하고 있습니다. 이와 관련된 [코멘트](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1162/changes#r2659813106)는 아래에 남겨두었습니다! 

아키텍처에 대한 제 생각을 남겨볼게요! 
`NotificationService`를 인터페이스가 아닌 클래스로 구현하고, `FcmNotificationService`에서 수행하던 비즈니스 로직을 옮기는 것은 어떨까요? 그리고 `NotificatioinSender` 인터페이스를 두고, `FcmNotificationService`를 구현체로 두면 책임을 분리할 수 있을 것 같습니다! 

조금 과할 수 있지만..  우선 말씀드려 봅니다! 

그림이 안떠오를 수 있으니 아키텍처 남길게요!

**현재 아키텍처**
<img width="866" height="546" alt="Image" src="https://github.com/user-attachments/assets/edc39ae7-0022-4f71-9437-75399338768e" />

**변경 아키텍처**
<img width="888" height="666" alt="Image" src="https://github.com/user-attachments/assets/f9d86821-141b-4e7e-aeb0-aa851d401378" />

jye326 replied:
반영했습니다~ 좋은데요~?

dye0p reviewed (commented):
추가로 코멘트 남겼습니다! 신경써서 구현하신게 느껴져서 리뷰하기 수월했습니다! 수고하셨습니다~

dye0p commented on backend/fittoring/src/main/java/fittoring/infrastructure/FcmNotificationService.java:
```diff
@@ -0,0 +1,47 @@
+
+@RequiredArgsConstructor
+@Service
+public class FcmNotificationService implements NotificationService {
```
p2: 아래 [코멘트](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1162#discussion_r2659813106)의 연장선입니다. 해당 클래스는 인프라 계층의 알림 전송을 담당하는 객체이므로 "비즈니즈 로직을 수행한다" 라는 네이밍의 Service보단, `FcmNofiticationSender` 같은 네이밍이 어떨까요?

jye326 replied:
`FcmNofiticationSender` 분리했습니다~

jye326 committed:
refactor: Device 모델 리팩토링 및 FCM 관련 로직 개선

- Device 모델의 token 및 isEnabled 필드 제거 후 hardwareId와 pushToken 추가
- N:1 관계 설정을 위해 Member 참조 필드 도입
- 회원당 기기 등록 제한 로직 및 예외(TooManyDeviceException) 추가
- FcmService의 upsertFcmToken 메서드 개선으로 디바이스 중복 처리 로직 변경
- FcmNotificationService 수정으로 멤버에 등록된 모든 기기에 알림 전송 가능
- 관련 SQL 스키마와 테스트 코드 업데이트 및 개선

jye326 committed:
refactor: FCM 관련 로직 개선 및 Device 모델 변경

- FcmService 제거 및 NotificationService 생성으로 역할 구분 강화
- Device 모델에 isPushEnabled 필드 추가 및 관련 메서드 구현
- FCM 발송 책임을 NotificationSender 인터페이스로 분리
- hardwareId 및 pushToken 통한 디바이스 관리 로직 정리
- 회원과 관련된 모든 FCM 알림 로직과 예외 처리 개선
- 관련 SQL 스키마, 서비스, 컨트롤러, 테스트 코드 업데이트

jye326 committed:
refactor: FCM 토큰 업서트 메서드 명칭 수정 및 관련 테스트 코드 변경

- upsertFcmToken → upsertPushToken으로 메서드 명칭 변경
- NotificationService 및 NotificationController 수정
- 관련 테스트 메서드 명칭 및 호출 변경

jye326 committed:
refactor: FcmTokenUpsertRequest를 PushTokenUpsertRequest로 리네이밍 및 관련 코드 수정

- FcmTokenUpsertRequest → PushTokenUpsertRequest로 이름 변경
- pushToken 필드로 명칭 수정
- NotificationController 및 관련 테스트 코드 업데이트

dye0p reviewed (approved):
코멘트 반영이 어느정도 된 것 같아요~ 승인합니다! 고생하셨습니다 :)
